protocol/messenger_mention.go

```diff
@@ -0,0 +1,1115 @@
+	regex := regexp.MustCompile(regexStr)
+	lCaseText := strings.ToLower(text)
+	return regex.MatchString(lCaseText)
+}
+
+func MatchMention(text string, users map[string]*MentionableUser, mentionKeyIdx int) *MentionableUser {
+	return matchMention(text, users, mentionKeyIdx, mentionKeyIdx+1, nil)
+}
+
+func matchMention(text string, users map[string]*MentionableUser, mentionKeyIdx int, nextWordIdx int, words []string) *MentionableUser {
+	if word := wordRegex.FindString(text[nextWordIdx:]); word != "" {
+		newWords := append(words, word)
+
+		t := strings.TrimSpace(strings.ToLower(strings.Join(newWords, "")))
+		searchedText := t
+		if lastChar := len(t) - 1; lastChar >= 0 && endingCharsRegex.MatchString(t[lastChar:]) {
+			searchedText = t[:lastChar]
+		}
+
+		userSuggestions := getUserSuggestions(users, searchedText, -1)
+		fmt.Printf("searchedText: %v, user suggestions: %v\n", searchedText, userSuggestions)
+		userSuggestionsCnt := len(userSuggestions)
+		switch {
+		case userSuggestionsCnt == 0:
+			return nil
+		case userSuggestionsCnt == 1:
+			user := getFirstUser(userSuggestions)
+			if isMentioned(user, text[mentionKeyIdx+1:]) {
+				return user
+			}
+		case userSuggestionsCnt > 1:
+			wordLen := len(word)
+			textLen := len(text)
+			nextWordStart := nextWordIdx + wordLen
+			if textLen > nextWordStart {
+				return matchMention(text, users, mentionKeyIdx, nextWordStart, newWords)
+			}
+		}
+	}
+	return nil
+}
+
+func getFirstUser(userSuggestions map[string]*MentionableUser) *MentionableUser {
+	for _, user := range userSuggestions {
+		return user
+	}
+	return nil
+}
+
+func ReplaceMentions(text string, users map[string]*MentionableUser) string {
+	idxs := getAtSigns(text)
+	return replaceMentions(text, users, idxs, 0)
+}
+
+func replaceMentions(text string, users map[string]*MentionableUser, idxs []int, diff int) string {
+	if strings.TrimSpace(text) == "" || len(idxs) == 0 {
+		return text
+	}
+
+	mentionKeyIdx := idxs[0] - diff
+
+	if len(users) == 0 {
+		return text
+	}
+
+	matchUser := MatchMention(text, users, mentionKeyIdx)
+	if matchUser == nil {
+		return replaceMentions(text, users, idxs[1:], diff)
+	}
+
+	newText := text[:mentionKeyIdx+1] + matchUser.ID + text[mentionKeyIdx+1+len(matchUser.Match):]
+	newDiff := diff + len(text) - len(newText)
+
+	return replaceMentions(newText, users, idxs[1:], newDiff)
+}
+
+func addSearchablePhrases(user *MentionableUser) *MentionableUser {
+	if !user.Blocked {
+		searchablePhrases := []string{user.PrimaryName(), user.SecondaryName()}
+		for _, s := range searchablePhrases {
+			if s != "" {
+				newWords := []string{s}
+				newWords = append(newWords, strings.Split(s, " ")[1:]...)
+				user.searchablePhrases = append(user.searchablePhrases, newWords...)
+			}
+		}
+		return user
+	}
+	return nil
+}
+
+type AtIndexEntry struct {
+	From    int
+	To      int
+	Checked bool
+
+	Mentioned bool // corresponding :mention?
```

Comment:
These 2 comments seem superfluous and related to the Clojure code. What does this comment mean to somebody working in the Desktop client?

Reply:
removed